Use extended mouse buttons for back / forward actions (GH5728)

neilcsmith-net opened the pull request:
This PR adds support for binding the back and forward buttons on a mouse to the back and forward code navigation actions, as requested in #5728 (and https://bz.apache.org/netbeans/show_bug.cgi?id=70961 )

This PR also allows binding extended mouse buttons, with or without modifiers, to any action in the keymap editor.

Both mouse buttons 4 and 6 are bound to back, and 5 and 7 to forwards, due to platform differences.  Triggering a dev build so we can get some testing.

Yes, treating mouse buttons as keycodes might seem slightly odd, but as we're already doing that for the mouse wheel editor zoom actions, which seems even odder ...

EDIT : screenshot of the updated special key panel, accessible from the dropdown when editing shortcuts 

![Screenshot from 2023-09-12 09-11-18](https://github.com/apache/netbeans/assets/3975960/5b15cdcf-c749-487a-a31c-2ce4a0334641)

mbien commented:
took me a while to realize that the "..." button transforms into a dropdown button which opens the special buttons popup and that mouse buttons don't automatically resolve into a value when you press edit.

But I think it works. I mapped `MOUSE_BUTTON6` and `7` which is forward/back on my mouse to comment and uncomment and it worked. Tried the same with 4/5 which are mouse wheel left/right and it worked too. Using logitech G502 which has a few buttons.

(what confused me initially was that `xev -event button | grep button` uses different button numbers than java does, so either the linux distribution or java must map them somehow)

btw what is the forward/back action?

neilcsmith-net commented:
> took me a while to realize that the "..." button transforms into a dropdown button which opens the special buttons popup and that mouse buttons don't automatically resolve into a value when you press edit.

Yes. Have added a screenshot of this to the description.

> (what confused me initially was that `xev -event button | grep button` uses different button numbers than java does, so either the linux distribution or java must map them somehow)

Confused me for a while with `xev` too.  I realised that X11 is mapping the mouse wheel up/down as buttons 4 & 5, so these are skipped in Java's button numbering.

Think it's explained by the code at https://github.com/openjdk/jdk/blob/master/src/java.desktop/unix/classes/sun/awt/X11/XWindow.java#L731
 
> btw what is the forward/back action?

An editor action, also in the editor toolbar.  Try `CTRL-click` on a method to open another method in the same or a different file.  Now try the back button (or in NB19 `ALT-LEFT`).  It's a multi-level navigation history - very useful when exploring code across multiple files.

mbien commented:
> An editor action, also in the editor toolbar. ... (or in NB19 `ALT-LEFT`).

right, very useful, Just to explain: I have been using this action all the time, just that when I moved the mouse over the toolbar button it showed the primary hotkey in the tooltip, so my brain went yesterday right away to: "well, must be some other action then". But its the same and the mouse is registered as second input source -> awesome.

neilcsmith-net commented:
I'll merge first thing next week unless anyone has any objections / comments?

mbien reviewed (approved):
code looks good too - thanks for documenting the important bits

neilcsmith-net commented on platform/openide.util.ui/src/org/openide/util/Utilities.java:
```diff
@@ -790,6 +790,13 @@ private static synchronized NamesAndValues initNameAndValues() {
         values.put(0x290,"MOUSE_WHEEL_UP");
         values.put(0x291,"MOUSE_WHEEL_DOWN");
 
+        for (int button = 4; button < 10; button++) {
```
I deliberately used the same "magic constants" approach that is already in use for `0x290` and `0x291` in the editor.  If we do want to move to an API, then these should be functions, as the values might have to change in future.  And we should look at doing the same for the existing editor support.

neilcsmith-net commented on platform/core.windows/src/org/netbeans/core/windows/ShortcutAndMenuKeyEventProcessor.java:
```diff
@@ -297,6 +311,71 @@ private boolean processShortcut(KeyEvent ev) {
+    private void processMouseEvent(MouseEvent mev) {
+        if (mev.getID() != MouseEvent.MOUSE_PRESSED
+                || mev.getButton() <= MouseEvent.BUTTON3
+                || mev.getButton() > 10
```
I could have a look at adding to Utilities as well.

neilcsmith-net replied:
Actually, seeing these two code fragments together shows they're different ranges anyway! :roll_eyes:

I'll look at adding something like `Utilities.isMouseKeyCode(int keyCode);` which can be used in the MainMenuAction filter as well.

neilcsmith-net commented on platform/openide.util.ui/src/org/openide/util/Utilities.java:
```diff
@@ -790,6 +790,13 @@ private static synchronized NamesAndValues initNameAndValues() {
+        for (int button = 4; button < 10; button++) {
+            String name = "MOUSE_BUTTON" + button; // NOI18N
+            int code = 0x292 + (button - 1);
+            names.put(name, code);
```
It's my understanding that check was added because of those keys being added to the JDK?  If anything we have the opposite problem and should check if the keycode is defined, but that applies to the existing mouse wheel pseudocodes already.

neilcsmith-net commented:
@sdedic thanks for the review. I've added API methods for querying pseudo-keycodes for mouse events, and switched to them in other modules, including the existing editor zoom action.

I've kept away from introducing constants as we may in future need to consider changing the values used underneath.

One remaining use of magic constants of a sort is the names used in the UI.

mbien reviewed (approved):
update looks good
